keps/sig-scheduling/3022-min-domains-in-pod-topology-spread/README.md

```diff
@@ -0,0 +1,506 @@
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+A new field `minDomains` is introduced to `PodSpec.TopologySpreadConstraint[*]` to limit
+the minimum number of topology domains. It functions in a mandatory or best-efforts manner,
+depending on the type of `WhenUnsatisfiable`.
+
+## Motivation
+
+Pod Topology Spread has [`maxSkew` parameter](https://github.com/kubernetes/enhancements/tree/11a976c74e1358efccf251d4c7611d05ce27feb3/keps/sig-scheduling/895-pod-topology-spread#maxskew), which control the degree to which Pods may be unevenly distributed.
+But, there isn't a way to control the number of domains over which we should spread.
+In some cases, users want to limit the number of domains for the cluster autoscaler and force spreading Pods over a minimum number of domains.
+
+### Goals
+
+- Users can specify `minDomains` to limit the number of domains.
+- Users can use it as a mandatory requirement with `WhenUnsatisfiable=DoNotSchedule`.
+- Users can use it as a best-efforts manner with `WhenUnsatisfiable=ScheduleAnyway`.
+
+### Non-Goals
+
+- Add new field to limit the maximum number of topology domains.
+
+## Proposal
+
+### User Story
+
+I am using cluster autoscaler and I want to force spreading a deployment over at least 5 Nodes.
+
+## Design Details
+
+Users can define a minimum number of domains with `minDomains` parameter.
+
+Pod Topology Spread has the semantics of "global minimum", which means the minimum number of pods that match the label selector in a topology domain.
+
+When the number of domains that have matching Pods is less than `minDomains`,
+Pod Topology Spread treats "global minimum" as 0.
+
+As the result,
+
+- when `whenUnsatisfiable` equals `DoNotSchedule`, scheduler doesn't schedule a matching Pod to Nodes on the domains that have the same number of matching Pods as `maxSkew`.
+- when `whenUnsatisfiable` equals `ScheduleAnyway`, scheduler prefers Nodes on the domains that don't have the same number of matching Pods as `maxSkew`.
+
+### API
+
+New optional parameter called `MinDomains` is introduced to `PodSpec.TopologySpreadConstraint[*]`.
+
+```go
+type TopologySpreadConstraint struct {
+......
+  // When the number of domains that have matching Pods is less than `minDomains`,
+  // Pod Topology Spread treats "global minimum" as 0.
+  // As the result,
+  // - when `whenUnsatisfiable` equals `DoNotSchedule`,
+  //   scheduler doesn't schedule a matching Pod to Nodes on the domains that have the same number of matching Pods as `maxSkew`.
+  // - when `whenUnsatisfiable` equals `ScheduleAnyway`,
+  //   scheduler prefers Nodes on the domains that don't have the same number of matching Pods as `maxSkew`.
+  // +optional
+  MinDomains *int32
```

Comment:
What is the default value? Probably zero. Please specify.

Reply:
Yes, it's 0. 
Sure, I'll add the description.